x-pack/plugins/infra/public/pages/logs/settings/form_elements.tsx

```diff
@@ -0,0 +1,243 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import equal from 'fast-deep-equal';
+import { useCallback, useMemo, useState } from 'react';
+import { useAsync } from 'react-use';
+import { ObjectEntries } from '../../../../common/utility_types';
+import { ChildFormValidationError, GenericValidationError } from './validation_errors';
+
+const unsetValue = Symbol('unset form value');
+
+type ValueUpdater<Value> = (updater: (previousValue: Value) => Value) => void;
+
+export interface FormElement<Value, InvalidReason> {
+  initialValue: Value;
+  isDirty: boolean;
+  resetValue: () => void;
+  updateValue: ValueUpdater<Value>;
+  validity: FormElementValidity<InvalidReason | GenericValidationError>;
+  value: Value;
+}
+
+type FormElementMap<FormValues extends {}> = {
+  [formElementName in keyof FormValues]: FormElement<FormValues[formElementName], any>;
+};
+
+export interface CompositeFormElement<CompositeValue extends {}, InvalidReason>
+  extends FormElement<CompositeValue, InvalidReason | ChildFormValidationError> {
+  childFormElements: FormElementMap<CompositeValue>;
+}
+
+export type FormElementValidity<InvalidReason> =
+  | { validity: 'valid' }
+  | { validity: 'invalid'; reasons: InvalidReason[] }
+  | { validity: 'pending' };
+
+export const useFormElement = <Value, InvalidReason>({
+  initialValue,
+  validate,
+}: {
+  initialValue: Value;
+  validate?: (value: Value) => Promise<InvalidReason[]>;
+}): FormElement<Value, InvalidReason> => {
+  const [changedValue, setChangedValue] = useState<Value | typeof unsetValue>(unsetValue);
+
+  const value = changedValue !== unsetValue ? changedValue : initialValue;
+
+  const updateValue = useCallback<ValueUpdater<Value>>(
+    (updater) =>
+      setChangedValue((previousValue) =>
+        previousValue === unsetValue ? updater(initialValue) : updater(previousValue)
+      ),
+    [initialValue]
+  );
+
+  const resetValue = useCallback(() => setChangedValue(unsetValue), []);
+
+  const isDirty = useMemo(() => !equal(value, initialValue), [value, initialValue]);
+
+  const validity = useValidity(value, validate);
+
+  return useMemo(
+    () => ({
+      initialValue,
+      isDirty,
+      resetValue,
+      updateValue,
+      validity,
+      value,
+    }),
+    [initialValue, isDirty, resetValue, updateValue, validity, value]
+  );
+};
+
+export const useCompositeFormElement = <FormValues extends {}, InvalidReason>({
+  childFormElements,
+  validate,
+}: {
+  childFormElements: FormElementMap<FormValues>;
+  validate?: (values: FormValues) => Promise<InvalidReason[]>;
+}): CompositeFormElement<FormValues, InvalidReason> => {
+  const childFormElementEntries = useMemo(
```

Comment:
:information_source: This serves as the key for many memoization operations below.